Feature #70 - DummyDataMaker 변경

chanwooo opened the pull request:
close #70

chanwooo committed:
#46-reviewService
getlist -> getListByStore, getListByUser 분리

chanwooo committed:
#46-reviewService
- delete요청한 유저가 review의 소유자인지 확인하는로직 추가

chanwooo committed:
#46-reviewController 작성

chanwooo committed:
#70-DummyDataMaker with id 추가

chanwooo committed:
#70-DummyDataMaker 사용하는부분들 수정

f-lab-trey commented on src/main/java/flab/quing/review/ReviewController.java:
```diff
@@ -0,0 +1,90 @@
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/api/v1/review")
+public class ReviewController {
```
아래 PR과 겹치는 내용인 거 같습니다.

f-lab-trey commented on src/test/java/flab/quing/DummyDataMaker.java:
```diff
@@ -11,6 +11,7 @@
 public class DummyDataMaker {
 
     static long userCount, storeCount, waitingCount, reviewCount;
+    static boolean withId = true;
```
이 상태를 유지하지 않고는 방법이 없을까요? DummyDataMaker는 많은 테스트 클래스들이 공유하는데, 이 상태를 유지하는 것은 `setWithId()`를 호출하느냐 아니냐에 따라 테스트를 실패하게 만들 수도 있어보입니다.

chanwooo replied:
컨텍스트를 올려서 테스트하는경우에는 id가 필요없으니 before()에서 setWithId(false)를하고 after()에서 true로 다시 돌리는 작업을 하고 있어요. 컨텍스트 올리지 않는 테스트에서는 true인 채로 유지되겠지만 after before를 잘못하면 엄한 테스트가 터질수도있다는것 인지하고있습니다.
static으로 만들어서 생성없이 바로 쓰면 편하겠다 생각하고 이렇게 만들었는데 자꾸 다른테스트와 공유하는부분이 생기다보니
각 테스트에서 그냥 생성해서 쓰는게 더 낫지않았을까 하는 생각도 이제와서 드네요..

f-lab-trey replied:
DummyDataMaker를 Builder 형식으로 만들어서 `DummyDataMaker.withId(true)....` 이렇게 쓸 수 있으면 더 좋을 거 같습니다. boolean 값을 클래스 변수로 갖고 있다가 사용하는 것은 나중에 버그를 유발할 가능성이 매우 높습니다.

chanwooo replied:
`DummyDataMaker dummyDataMaker = DummyDataMaker.builder().withId(true).build();`
이런식으로 각 테스트마다 만들어서 사용하도록 변경하였습니다.

chanwooo committed:
Merge branch 'develop' into feature-#70

# Conflicts:
#	src/main/java/flab/quing/review/ReviewController.java

chanwooo committed:
#70 - DummyDataMaker 수정
- 객체 생성하는 메서드들 non static으로 변경
- withId를 설정하는부분 생성자, 빌더로 변경
- StoreManager 생성 추가

chanwooo committed:
#70 - 테스트코드 수정
- 변경된 dummyDataMaker에 따라서 객체 생성하고 생성된 객체의 생성메서드 사용

chanwooo commented:
`DummyDataMaker dummyDataMaker = DummyDataMaker.builder().withId(true).build();`
이런식으로 각 테스트마다 만들어서 사용하도록 변경하였습니다.

f-lab-trey commented on src/test/java/flab/quing/DummyDataMaker.java:
```diff
@@ -2,44 +2,46 @@
+    long userCount, storeCount, waitingCount, reviewCount, storeManagerCount;
 
-    public static void init() {
+    public void init() {
```
꼭 Builder의 `init()`을 외부에서 별도로 호출해줘야 할까요? 통상적으로 Builder들은 이런 과정 없이 사용할 수 있었던 거 같습니다.

chanwooo replied:
builder의 init이 아니라 `@Transactional`이 달린 테스트처럼 테스트케이스 하나를 실행하고
초기화되어야할때 호출하려고 만든 메서드입니다.

f-lab-trey commented on src/test/java/flab/quing/review/ReviewServiceImplTest.java:
```diff
@@ -39,10 +39,11 @@ class ReviewServiceImplTest {
     @BeforeEach
     void init() {
-        DummyDataMaker.init();
+        dummyDataMaker.init();
```
[q]
왜 일부 테스트 클래스에서만 `init()`이 호출될까요?

chanwooo replied:
`.init()`을 호출하는 test들이 entity의 id를 검증하는 케이스가 있어서 쓰게된거같은데 
id는 어떤값이든 나올수있는것이고..여기에 의존하면 안되는것이고..그러니 없애고 테스트를 수정하는게 맞겠네요.

그럼 init이란 메서드도 필요 없어질거같아요. id를 검증하는 테스트를 수정해보겠습니다~!

chanwooo committed:
#70 - DummyDataMaker init() 제거

chanwooo committed:
#70 - id를 상수로 직접 입력한것들 테스트용으로 만든 객체의 id를 이용하도록 변경
